scriptworker/gpg.py

```diff
@@ -1113,25 +1131,337 @@ def rebuild_gpg_home_signed(context, real_gpg_home, my_pub_key_path,
+        line = await proc.stdout.readline()
+        if line:
+            output += line.decode('utf-8')
+        else:
+            break
+    verify_signed_git_commit_output(output)
+
+
+# build gpg homedirs from repo {{{1
+async def build_gpg_homedirs_from_repo(
+    context, basedir=None, verify_function=verify_signed_git_commit,
+    flat_function=rebuild_gpg_home_flat, signed_function=rebuild_gpg_home_signed,
+):
+    """Build gpg homedirs in `basedir`, from the context-defined git repo.
+
+    Args:
+        context (scriptworker.context.Context): the scriptworker context.
+        basedir (str, optional): the path to the base directory to create the
+            gpg homedirs in.  This directory will be wiped if it exists.
+            If None, use `context.config['base_gpg_home_dir']`.  Defaults to None.
+
+    Returns:
+        str: on success.
+
+    Raises:
+        ScriptWorkerGPGException: on rebuild exception.
+    """
+    basedir = basedir or context.config['base_gpg_home_dir']
+    repo_path = context.config['git_key_repo_dir']
+    lockfile = os.path.join(basedir, ".lock")
+    if os.path.exists(lockfile):
+        log.warning("Skipping build_gpg_homedirs_from_repo: lockfile {} exists!".format(lockfile))
+        return
+    rm(basedir)
+    makedirs(basedir)
+    try:
+        # create lockfile
+        with open(lockfile, "w") as fh:
+            print(str(arrow.utcnow().timestamp), file=fh, end="")
+        # verify our input.  Hardcoding the check before importing, as opposed
+        # to expecting something else to run the check for us.
+        await verify_function(context)
+        # create gpg homedirs
+        tasks = []
+        for worker_class, worker_config in context.cot_config['gpg_homedirs'].items():
+            source_path = os.path.join(repo_path, worker_class)
+            real_gpg_home = os.path.join(basedir, worker_class)
+            my_pub_key_path = context.cot_config['pubkey_path']
+            my_priv_key_path = context.cot_config['privkey_path']
+            if worker_config['type'] == 'flat':
+                tasks.append(asyncio.ensure_future(
+                    flat_function(
+                        context, real_gpg_home, my_pub_key_path, my_priv_key_path,
+                        source_path, ignore_suffixes=worker_config['ignore_suffixes']
+                    )
+                ))
+            else:
+                trusted_path = os.path.join(source_path, "trusted")
+                untrusted_path = os.path.join(source_path, "valid")
+                tasks.append(asyncio.ensure_future(
+                    signed_function(
+                        context, real_gpg_home, my_pub_key_path, my_priv_key_path,
+                        trusted_path, untrusted_path=untrusted_path,
+                        ignore_suffixes=worker_config['ignore_suffixes']
+                    )
+                ))
+            await raise_future_exceptions(tasks)
+    finally:
+        # rm lockfile
```

Comment:
+1. Guessing this was a todo item or more complex logic at some point, not sure.  Removed the comment.